assets/app/scripts/controllers/edit/buildConfig.js

```diff
@@ -0,0 +1,312 @@
+          $scope.pushTo.tags[name] = tagList;
+        });
+        if (selectFirstOption) {
+          $scope.options.pickedPushToImageStream = $scope.pushTo.imageStreams[0];
+          $scope.clearSelectedOutputTag();
+        }
+      });
+    }
+
+    $scope.clearSelectedOutputTag = function() {
+      var tags = $scope.pushTo.tags[$scope.options.pickedPushToImageStream];
+      $scope.options.pickedPushToImageStreamTag = _.find(tags, function(tag) { return tag == "latest" }) || tags[0] || "";
+    }
+
+    $scope.save = function() {
+      $scope.disableInputs = true;
+      // Update Configuration
+      $filter('buildStrategy')($scope.updatedBuildConfig).forcePull = $scope.options.forcePull;
+      if ($scope.updatedBuildConfig.spec.strategy.type === "Docker") {
+        $filter('buildStrategy')($scope.updatedBuildConfig).noCache = $scope.options.noCache;
+      }
+
+      var from = {};
+      if ($scope.options.pickedBuildFromType === "ImageStreamTag") {
+        from = {
+          kind: $scope.options.pickedBuildFromType,
+          namespace: $scope.options.pickedBuildFromNamespace,
+          name: $scope.options.pickedBuildFromImageStream + ":" + $scope.options.pickedBuildFromImageStreamTag
+        };
+      } else if ($scope.options.pickedBuildFromType === "DockerImage") {
+        from = {
+          kind: $scope.options.pickedBuildFromType,
+          name: $scope.options.pickedBuildFromDockerImage
+        };
+      }
+      $filter('buildStrategy')($scope.updatedBuildConfig).from = from;
+
+      var to = {
+        kind: "ImageStreamTag",
+        namespace: $scope.options.pickedPushToNamespace,
+        name: $scope.options.pickedPushToImageStream + ":" + $scope.options.pickedPushToImageStreamTag
+      };
+      $scope.updatedBuildConfig.spec.output.to = to;
+
+      // Update envVars
+      var updateEnvVars = [];
+      angular.forEach($scope.envVars, function(v, k) {
+        var env = {
+          name: k,
+          value: v
+        };
+        updateEnvVars.push(env);
+      });
+      $filter('buildStrategy')($scope.updatedBuildConfig).env = updateEnvVars
+
+      // Update triggers
+      var triggers = [];
+      if ($scope.triggers.webhook) {
+        var webhook = _.find($scope.buildConfig.spec.triggers, function(obj) { return obj.type == "GitHub" })
+        if (!webhook) {
+          webhook =  {
+            github: {
+              secret: ApplicationGenerator._generateSecret()
+            },
+            type: "GitHub"
+          };
+        }
+        triggers.push(webhook);
+      }
+      triggers.push(_.find($scope.buildConfig.spec.triggers, function(obj) { return obj.type == "Generic" }));
+
+      if ($scope.triggers.imageChange) {
+        var imageChange = _.find($scope.buildConfig.spec.triggers, function(obj) { return obj.type == "ImageChange" });
+        if (!imageChange) {
+          triggers.push({
+            imageChange: {},
+            type: "ImageChange"
+          });
+        }
```

Comment:
Existing image change triggers aren't going to be added back to the triggers array, so we could lose data on edit here.

How about

```
    var imageChange = _.find($scope.buildConfig.spec.triggers, function(obj) { return obj.type == "ImageChange" });
    if (!imageChange) {
      imageChange = {
        imageChange: {},
        type: "ImageChange"
      };
    }
    triggers.push(imageChange);
 ```
 
 or if you use `_.filter`
 
```
    var imageChangeTriggers = _.filter($scope.buildConfig.spec.triggers, function(obj) { return obj.type == "ImageChange" });
    if (!imageChangeTriggers.length) {
      imageChangeTriggers.push({
        imageChange: {},
        type: "ImageChange"
      });
    }
    triggers.concat(imageChangeTriggers);
 ```
 
 (untested)